### PR TITLE
Make gradient accumulation work with dispatched dataloaders

### PR DESCRIPTION
**Pull request by muellerzr**
This PR implements gradient accumulation support for dispatched dataloaders and rewrites the tests for testing all combinations of dispatched dataloaders + split batches

### CODE REVIEW

File: src/accelerate/test_utils/scripts/test_sync.py [hunk 1]
```diff
@@ -229,12 +238,28 @@ def main():
...
     test_gradient_accumulation_with_opt_and_scheduler()
+    if state.distributed_type == DistributedType.MULTI_GPU:
+        for split_batch in [True, False]:
+            for dispatch_batches in [True, False]:
+                if not split_batch and not dispatch_batches:
+                    continue
+                if state.local_process_index == 0:
+                    print(
+                        f"**Test `accumulate` gradient accumulation with optimizer and scheduler, `split_batches={split_batch}` and `dispatch_batches={dispatch_batches}`**"
+                    )
```
Comment (sgugger) [inline src/accelerate/test_utils/scripts/test_sync.py:261]: Should be split in several lines to stay under the max char len of 119.
